crates/red_knot_python_semantic/src/types/builder.rs

```diff
@@ -216,74 +216,139 @@ impl<'db> InnerIntersectionBuilder<'db> {
     }
 
     /// Adds a positive type to this intersection.
-    fn add_positive(&mut self, db: &'db dyn Db, ty: Type<'db>) {
+    fn add_positive(&mut self, db: &'db dyn Db, new_positive: Type<'db>) {
         // TODO `Any`/`Unknown`/`Todo` actually should not self-cancel
-        match ty {
-            Type::Intersection(inter) => {
-                let pos = inter.positive(db);
-                let neg = inter.negative(db);
-                self.positive.extend(pos.difference(&self.negative));
-                self.negative.extend(neg.difference(&self.positive));
-                self.positive.retain(|elem| !neg.contains(elem));
-                self.negative.retain(|elem| !pos.contains(elem));
```

Comment:
This was certainly more efficient before. There are certainly ways to improve performance of `add_positive`/`add_negative`. Is it something I should look into now, or is it okay to postpone that to the point were the runtime of intersection-type-simplification becomes a problem (if ever).

Reply:
> we'll probably create lots of intersections once our model is more fleshed out

but will we also create *large* intersections? Since we have this eager approach now where we simplify immediately, what would be an example for a large intersection type (>= 3 elements)?

Reply:
Good question! Not really sure how this is going to play out, in all honesty. Possibly @carljm has a better answer for you here :-)